### PR TITLE
Rework PWM module to follow `Extended API Guideline` and update documents

### PR DESCRIPTION
**Pull request by hs0225**
- Tested on raspberryPi3, stm32f4-discovery

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

File: docs/api/IoT.js-API-PWM.md [hunk 1]
```diff
@@ -1,89 +1,247 @@
...
+| pwmpin.setEnableSync | O | O | O |
+| pwmpin.close | O | O | O |
+| pwmpin.closeSync | O | O | O |
+
+
+## Contents
+* [Pin](#pin)
+* [PWM](#pwm)
+  * [Constructor](#pwm-constructor)
+    * [`new ADC()`](#pwm-new)
```
Comment (chokobole) [inline docs/api/IoT.js-API-PWM.md]: new PWM()

File: docs/api/IoT.js-API-PWM.md [hunk 2]
```diff
@@ -1,89 +1,247 @@
...
+  * [Prototype methods](#pwm-prototype-methods)
+    * [`adc.open(configuration[, callback])`](#pwm-open)
```
Comment (chokobole) [inline docs/api/IoT.js-API-PWM.md]: pwm.open

File: docs/api/IoT.js-API-PWM.md [hunk 3]
```diff
@@ -1,89 +1,247 @@
...
 
 
-### Constructor
+## Class: PWM <a name="pwm"></a>
+
+
+## Constructor <a name="pwm-constructor"></a>
+
+
+### new PWM <a name="pwm-new"></a>
```
Comment (chokobole) [inline docs/api/IoT.js-API-PWM.md]: please embrace expression with ``

File: src/module/iotjs_module_pwm.c [hunk 4]
```diff
@@ -71,232 +100,278 @@ iotjs_pwm_reqdata_t* iotjs_pwm_reqwrap_data(THIS) {
...
-  iotjs_pwm_reqdata_t* req_data = iotjs_pwm_reqwrap_data(pwm_reqwrap);
-  req_data->pin = JHANDLER_GET_ARG(0, number);
-
-  const iotjs_jval_t* joptions = JHANDLER_GET_ARG(1, object);
-  if (joptions && iotjs_jval_is_object(joptions)) {
-    iotjs_jval_t jperiod =
-        iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PERIOD);
-    if (iotjs_jval_is_number(&jperiod))
-      req_data->period = iotjs_jval_as_number(&jperiod);
+iotjs_pwm_t* iotjs_pwm_instance_from_jval(const iotjs_jval_t* jpwm) {
```
Comment (chokobole) [inline src/module/iotjs_module_pwm.c]: If this can be `static function`, then please mark it with `static`

File: src/module/iotjs_module_pwm.c [hunk 5]
```diff
@@ -71,232 +100,278 @@ iotjs_pwm_reqdata_t* iotjs_pwm_reqwrap_data(THIS) {
...
 
-    iotjs_jval_t jduty_cycle =
-        iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_DUTYCYCLE);
-    if (iotjs_jval_is_number(&jduty_cycle))
-      req_data->duty_cycle = iotjs_jval_as_number(&jduty_cycle);
 
-    iotjs_jval_destroy(&jperiod);
-    iotjs_jval_destroy(&jduty_cycle);
-  }
+iotjs_pwm_t* iotjs_pwm_instance_from_reqwrap(THIS) {
```
Comment (chokobole) [inline src/module/iotjs_module_pwm.c]: ditto

File: src/module/iotjs_module_pwm.c [hunk 6]
```diff
@@ -71,232 +100,278 @@ iotjs_pwm_reqdata_t* iotjs_pwm_reqwrap_data(THIS) {
...
+  const iotjs_jval_t* jpwm = JHANDLER_GET_THIS(object);
+  iotjs_pwm_t* pwm = iotjs_pwm_create(jpwm);
+  IOTJS_ASSERT(pwm == iotjs_pwm_instance_from_jval(jpwm));
 
-  iotjs_pwm_set_pwmdata(jhandler, req_wrap);
+  const iotjs_jval_t* jconfiguration = JHANDLER_GET_ARG(0, object);
+  const iotjs_jval_t* jcallback = JHANDLER_GET_ARG(1, function);
 
-  PWM_ASYNC(Export);
+  // set configuration
```
Comment (chokobole) [inline src/module/iotjs_module_pwm.c]: please capitalize beginning letter

File: docs/api/IoT.js-API-PWM.md (hunk 1 above)
Comment (hs0225) [inline docs/api/IoT.js-API-PWM.md]: done.

File: docs/api/IoT.js-API-PWM.md (hunk 2 above)
Comment (hs0225) [inline docs/api/IoT.js-API-PWM.md]: done.

File: docs/api/IoT.js-API-PWM.md (hunk 3 above)
Comment (hs0225) [inline docs/api/IoT.js-API-PWM.md]: done.

File: src/module/iotjs_module_pwm.c (hunk 4 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.c]: done.

File: src/module/iotjs_module_pwm.c (hunk 5 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.c]: This is not a static function.

File: src/module/iotjs_module_pwm.c (hunk 6 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.c]: done.

File: docs/api/IoT.js-API-PWM.md [hunk 7]
```diff
@@ -1,89 +1,247 @@
...
 
 
-### Constructor
+## Class: PWM <a name="pwm"></a>
+
+
+## Constructor <a name="pwm-constructor"></a>
+
+
+### `new PWM` <a name="pwm-new"></a>
```
Comment (chokobole) [inline docs/api/IoT.js-API-PWM.md]: Please fix signature to `new PWM()` as it is typed at other place.

Comment (hs0225) [inline docs/api/IoT.js-API-PWM.md]: done.

**Comment by chokobole** [comment]
LGTM

**Comment by yichoi** [comment]
LGTM

**Pull request merged**